### PR TITLE
System profiles autodetection

### LINKED ISSUE #2459: add system profile auto detection based off RAM, vRAM, and system architecture

### PR DESCRIPTION
**Pull request by cdoern**
detect system profiles based on exact matches only. If a user is on an Apple M3 Pro system, we will match that to the name of the profile we write to disk.
The same goes for all Nvidia system profiles currently supported.

We are not doing best approximation here, at least not yet. The approximation proved to add confusion with training profiles. We also cover many more use cases with system profiles than we did with training profiles so odds are you should have some sort of direct match!

Here is what `ilab config init` looks like after these changes:

Screenshot from 2024-11-05 13-26-45

the interactive menu now only pops up if no direct profile matches are found. 

resolves #2459

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/config/init.py [hunk 1]
```diff
@@ -197,57 +181,147 @@ def walk_and_print_system_profiles() -> Config | None:
...
+    )
+    # the file is SYSTEM_PROFILE_DIR/arch_family_procesors[key][selection-1]
+    if 1 <= system_profile_selection <= len(arch_family_processors[key]):
+        file = os.path.join(
+            DEFAULTS.SYSTEM_PROFILE_DIR,
+            "/".join(
+                map(str, arch_family_processors[key][system_profile_selection - 1])
+            ),
+        )
+        click.echo(click.style(f"You selected: {file}", fg="green"))
```
Comment (nathan-weinberg) [inline src/instructlab/config/init.py:206]:
```suggestion
        click.secho(f"You selected: {file}", fg="green")
```

File: src/instructlab/config/init.py [hunk 2]
```diff
@@ -197,57 +181,147 @@ def walk_and_print_system_profiles() -> Config | None:
...
+        click.secho(
+            "Invalid selection. Please select a valid system profile option.",
+            fg="red",
+        )
+        raise click.exceptions.Exit(1)
+    return cfg
+
+
+# walk_and_print_system_profiles prints interactive prompts asking the user to choose
+# their hardware vendor and system profile
```
Comment (nathan-weinberg) [inline src/instructlab/config/init.py:222]: Think this comment can be removed given the below docstring is the same

File: src/instructlab/config/init.py [hunk 3]
```diff
@@ -197,57 +181,147 @@ def walk_and_print_system_profiles() -> Config | None:
...
+                file = os.path.join(DEFAULTS.SYSTEM_PROFILE_DIR, system_profile_file)
+            elif chip_name == " ".join(printed_arch_family_processor):
+                # else the whole thing needs to match, so `nvidia l4 x8` for example would be the chip we detect AND the arch family processor from the list of files
+                file = os.path.join(DEFAULTS.SYSTEM_PROFILE_DIR, system_profile_file)
+            if file is not None:
+                click.echo(
+                    click.style(
+                        f"We have detected {file} as an exact match for your system.",
+                        fg="green",
+                    )
```
Comment (nathan-weinberg) [inline src/instructlab/config/init.py:277]:
```suggestion
                click.secho(
                    f"We have detected {file} as an exact match for your system.",
                    fg="green",
                )
```

File: src/instructlab/configuration.py [hunk 4]
```diff
@@ -1189,6 +1189,7 @@ def read_and_create_system_profiles(profiles_dir: str, overwrite: bool) -> bool:
 
 # recreate_system_profiles writes all profile directories found in src/instructlab to disk
 # the location is ~/.local/share/instructlab/internal/system_profiles
+# pylint: disable=broad-exception-caught
```
Comment (nathan-weinberg) [inline src/instructlab/configuration.py:1192]: Why is this being added in isolation?

**Review by nathan-weinberg — COMMENTED** [review]
Will give it a second look after others review but here is a first pass

File: src/instructlab/configuration.py (hunk 4 above)
Comment (cdoern) [reply src/instructlab/configuration.py:1192]: leftover, removed. Thanks!

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,13 @@
+## v0.21
+
+### Breaking Changes
+
+* train-profiles have been deprecated and replaced with system-profiles. These profiles follow the format of the config file and apply to all commands. They live in `~/.local/share/instructlab/internal/system_profiles`
+
+### Features
+
+* `ilab` now support system profiles. These profiles apply entire configuration files tailored to specific hardware configurations. We support a set of auto-detected profiles for CPU enabled Linux machines, M-Series Apple Silicon Chips, and Nvidia GPUs. When you run `ilab config init`, one of these profiles should be selected for you. If there is not a direct match, a menu will be displayed allowing you to choose one.
```
Comment (alinaryan) [inline CHANGELOG.md:9]: nit: support->supports

**Review by alinaryan — APPROVED** [review]
Looks good! Just one nit. I'll follow your lead with the tests in my gaudi PR

File: src/instructlab/config/init.py [hunk 6]
```diff
@@ -357,3 +439,46 @@ def get_separator(text: str) -> str:
...
+                    if "model name" in line:
+                        # print(line)
+                        return (
+                            line.split(":")[-1].strip().lower().split(" ")[0],
+                            True,
+                            True,
+                        )
+        except FileNotFoundError:
+            return "Unsupported", True, True
+
```
Comment (jaideepr97) [inline src/instructlab/config/init.py:483]: can we add an else to inform users we only support linux/mac?

File: src/instructlab/config/init.py [hunk 7]
```diff
@@ -197,57 +181,155 @@ def walk_and_print_system_profiles() -> Config | None:
...
+    return cfg
+
+
+def walk_and_print_system_profiles(
+    is_linux: bool, is_cpu: bool, chip_name: str | None = None
+) -> Config | None:
+    """
+    walk_and_print_system_profiles prints interactive prompts asking the user to choose
+    their hardware vendor and system profile
+    """
```
Comment (jaideepr97) [inline src/instructlab/config/init.py:227]: please update the docstring to specify that this function doesn't just print the profiles, it also picks one and returns a config object for it

**Review by jaideepr97 — COMMENTED** [review]
thanks @cdoern 
few minor comments

File: src/instructlab/config/init.py (hunk 6 above)
Comment (cdoern) [reply src/instructlab/config/init.py:483]: sure! I can add a log warning

File: src/instructlab/config/init.py (hunk 7 above)
Comment (cdoern) [reply src/instructlab/config/init.py:227]: done

**Review by jaideepr97 — APPROVED** [review]

**Comment by cdoern** [comment]
tests keep hanging for me locally but rebasing seems to have fixed it so I re-pushed... weird

**Pull request merged**